weldx/asdf/file.py

```diff
@@ -0,0 +1,357 @@
+import pathlib
+from collections import UserDict
+from io import BytesIO, IOBase
+from typing import Optional, Protocol, Union, runtime_checkable, Mapping
+
+import asdf
+from asdf import AsdfFile
+
+from weldx.asdf import WeldxAsdfExtension, WeldxExtension
+
+#TODO: asdf internally checks interfaces sometimes by attribute, sometimes if it is an instance of io.IOBase etc.
+# so this is actually too broad!
+@runtime_checkable
+class SupportsFileReadOnly(Protocol):
+    """Type interface for read()."""
+
+    __slots__ = ()
+
+    def read(self):
+        raise NotImplementedError
+
+
+@runtime_checkable
+class SupportsFileReadWrite(Protocol):
+    """Type interface for read, write and seeking."""
+
+    __slots__ = ()
+
+    def read(self):
+        raise NotImplementedError
+
+    def write(self, *args):
+        raise NotImplementedError
+
+    def tell(self):
+        raise NotImplementedError
+
+    def seek(self, *args):
+        raise NotImplementedError
+
+
+types_file_like = Union[IOBase, SupportsFileReadOnly, SupportsFileReadWrite]
+types_path_and_file_like = Union[str, pathlib.Path, types_file_like]
+
+
+class WeldxFile(UserDict):
+    """This exposes an ASDF file as a dictionary like object.
+
+    Parameters
+    ----------
+    filename_or_file_like :
+        a path to a weldx file or file handle like to read/write data from.
+        If None is passed, an in-memory file will be created.
+    mode :
+        reading or reading/writing mode: "r" or "rw"
+    asdf_args :
+        see `asdf.open` for reference.
+    tree :
+        an optional dictionary to write to the file.
+    sync :
+        If True, the changes to file will be written upon closing this. This is only
+        relevant, if the file has been opened in write mode.
+    software_history_entry :
+        a optional dictionary which will be used to add history entries upon
+        modification of the file. It has to provide the following keys:
+        ("name", "author", "homepage", "version")
+
+    """
+
+    def __init__(
+        self,
+        filename_or_file_like: Union[None, str, pathlib.Path, types_file_like],
+        mode: str = "r",
+        asdf_args: Mapping = None,
+        tree: Mapping = None,
+        sync: bool = True,
+        software_history_entry: Mapping = None,
+    ):
+        # TODO: default asdf_args?
+        # e.g: asdf_args = {"copy_arrays": True}
+        if asdf_args is None:
+            asdf_args = {}
+
+        self._quality_standard = (
+            asdf_args["custom_schema"] if "custom_schema" in asdf_args else None
+        )
```

Comment:
```suggestion
        self._custom_schema = (
            asdf_args["custom_schema"] if "custom_schema" in asdf_args else None
        )
```
might get confusion with the "full" quality standards later so I suggest sticking to ASDF naming

Reply:
OK thanks, I rename it to custom_schema then.